feat: add `VZVmnetNetworkDeviceAttachment` support (macOS 26.0)

norio-nomura opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, please read our contributor guidelines: https://github.com/Code-Hex/vz/blob/master/CONTRIBUTING.md
2. Please create a new issue before creating this PR. However, You can continue it without creating issues if this PR fixes any documentations such as typo.
3. Do not send Pull Requests for large (150 ~ lines) code changes. If so, I am not motivated to review your code. Basically, I write the code.
-->

## feat: add `VZVmnetNetworkDeviceAttachment` support (macOS 26.0)
`VZVmnetNetworkDeviceAttachment` is an API that creates vmnet devices on VMs added in macOS 26.

see: https://developer.apple.com/documentation/virtualization/vzvmnetnetworkdeviceattachment?language=objc

It does not require the `com.apple.vm.networking` entitlement nor root privileges.
`HostMode` and `SharedMode` are supported.
In order for multiple VMs to communicate with each other in `SharedMode`, they must be started in the same executable and the same `VmnetNetwork` must be passed to `NewVmnetNetworkDeviceAttachment()` to create an attachment.

This change adds:
- `vz.VmnetNetworkDeviceAttachment` represents `VZVmnetNetworkDeviceAttachment` in Go
- `vmnet` package to use vmnet APIs that added on macOS 26.0
  - `Return` represents `vmnet_return_t` as `error`
    - `ErrSuccess`, `ErrFailure`, ...
  - `Mode` represents `operating_modes_t`
    - `HostMode`, `SharedMode`
  - `NetworkConfiguration` represents `vmnet_network_configuration_t`
  - `Network` represents `vmnet_network_ref`
  - `Interface` represents `interface_ref`
  - `*FileAdaptorForInterface`s support File Handle based network device APIs on QEMU, krunkit, and `vz.NewFileHandleNetworkDeviceAttachment`

- `xpc` package that providing `<xpc/xpc.h>` APIs to support implementing Mach service server/client to sharing serializations of `vmnet.Network` and file descriptors of `vmnet.*FileAdaptorForInterface`s

- `vz_test.TestVmnetSharedModeAllowsCommunicationBetweenMultipleVMs`
- `vz_test.TestVmnetSharedModeWithConfiguringIPv4`
- `vz_test.TestVmnetNetworkShareModeSharingOverXpc`
`TestVmnetNetworkShareModeSharingOverXpc` tests sharing `vmnet.Network` in `SharedMode` over XPC communication.
This test registers test executable as an Mach service and launches it using `launchctl`.
The launched Mach service provides `vmnet.Network` serialization to clients upon request, after booting
a VM using the provided `vmnet.Network` to ensure the network is functional on the server side.
The client boots VM using the provided `vmnet.Network` serialization.

**Edit: on macOS 26.2, "the same executable" restriction seems to be relaxed.**
  - `VZVmnetNetworkDeviceAttachment` seems to allow connecting to subnet created by a different executable.
  - `vmnet_interface_start_with_network` seems to allow connecting to subnet created by an executable at same path? (may changing CDHash not affect?)
  - I don't know how far the restrictions have been relaxed.

## Which issue(s) this PR fixes:

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Mentioned in https://github.com/Code-Hex/vz/issues/198#issuecomment-3301732174

<!--
This section can be blank.
-->

norio-nomura committed:
feat: add `VZVmnetNetworkDeviceAttachment` support (macOS 26.0)

`VZVmnetNetworkDeviceAttachment` is an API that creates vmnet devices on VMs added in macOS 26.

see: https://developer.apple.com/documentation/virtualization/vzvmnetnetworkdeviceattachment?language=objc

It does not require the `com.apple.vm.networking` entitlement nor root privileges.
`HostMode` and `SharedMode` are supported.
In order for multiple VMs to communicate with each other in `SharedMode`, they must be started in the same executable and the same `VmnetNetwork` must be passed to `NewVmnetNetworkDeviceAttachment()` to create an attachment.

This change adds:
- `vz.VmnetNetworkDeviceAttachment` represents `VZVmnetNetworkDeviceAttachment` in Go
- `vmnet` package to use vmnet APIs that added on macOS 26.0
  - `Return` represents `vmnet_return_t` as `error`
    - `ErrSuccess`, `ErrFailure`, ...
  - `Mode` represents `operating_modes_t`
    - `HostMode`, `SharedMode`
  - `NetworkConfiguration` represents `vmnet_network_configuration_t`
  - `Network` represents `vmnet_network_ref`

- `vz_test.TestVmnetSharedModeAllowsCommunicationBetweenMultipleVMs`
- `vz_test.TestVmnetSharedModeWithConfiguringIPv4`

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

AkihiroSuda commented on vmnet.go:
```diff
@@ -0,0 +1,419 @@
+const (
+	HostMode   VmnetMode = C.VMNET_HOST_MODE
+	SharedMode VmnetMode = C.VMNET_SHARED_MODE
+	// Deprecated: BridgedMode is not supported by NewVmnetNetworkConfiguration
```
Do you know why not supported?

norio-nomura replied:
I don't know. It's documented on:
https://developer.apple.com/documentation/vmnet/vmnet_network_configuration_create(_:_:)?language=objc
> Parameters
  mode
    Shared mode or host-only mode.

nirs replied:
Why BridgedMode is Deprecated? Can you link to the docs about this?

Since it is not supported we should not include it, marking it as Deprecated look wrong.

norio-nomura committed:
vmnet_test.go: Add `TestVmnetNetworkShareModeSharingOverXpc`

- Add `TestVmnetNetworkShareModeSharingOverXpc` to `vmnet_test.go`
`TestVmnetNetworkShareModeSharingOverXpc` tests sharing `vmnet.Network` in `SharedMode` over XPC communication.
This test registers test executable as an Mach service and launches it using `launchctl`.
The launched Mach service provides `vmnet.Network` serialization to clients upon request, after booting
a VM using the provided `vmnet.Network` to ensure the network is functional on the server side.
The client boots VM using the provided `vmnet.Network` serialization.

- Add `pkg/xpc` package that providing `<xpc/xpc.h>` APIs to support implementing Mach service server and client

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

norio-nomura commented:
> It seems that it can be reproduced in the unit test, so I will try to make a unit test.

Added unit test and `pkg/xpc`.

nirs reviewed (commented):
I did not review most of this change, just the stange part of about marking bridged mode as depracated.

norio-nomura commented on pkg/xpc/xpc.m:
```diff
@@ -0,0 +1,259 @@
+
+// xpc_object_t
+
+void *xpcRetain(void *o)
```
It seemed that it could be used if it was not exposed outside the package, but I wanted to expose it outside the package, so I used a method using `void *`.

AkihiroSuda commented:
> By the way, should I include the correction of macOS 13 runner's retirement from GitHub Actions in this PR?

Can be a separate PR

AkihiroSuda commented on virtualization_26.m:
```diff
@@ -0,0 +1,226 @@
+// see: https://developer.apple.com/documentation/vmnet/vmnet_network_configuration_create(_:_:)?language=objc
+void *VZVmnetNetworkConfigurationCreate(uint32_t mode, uint32_t *status)
+{
+#ifdef INCLUDE_TARGET_OSX_26
```
Can we test this on GHA with macos-26 ?

norio-nomura commented:
> > By the way, should I include the correction of macOS 13 runner's retirement from GitHub Actions in this PR?
> 
> Can be a separate PR

opened #208

norio-nomura committed:
Extend test timeouts

- .github/workflows/compile.yml: Extend test job's timeout from 3m to 5m
- Makefile: Extend test target's timeout from 2m to 4m

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

Code-Hex reviewed (commented):
Thank you very much for creating a great PR.

I don't think it's ready for review yet, but I'll review it in its current state.

1. When providing it as a library, it is not desirable to create a `pkg` directory and create detailed packages within it. Please remove the `pkg` directory and create `vmnet`, `xpc` directories at the same level as `network.go`, such as vz/network.go, vz/vmnet, vz/xpc
2. Similar to the above, create a directory for the `osversion` package under internal and move `osversion.go` it to use the `macOSAvailable` function in the `vmnet` package.
3. Overall, the code seems to have a tendency to overlap with existing code (does it support generative AI?). It would be appreciated if you could align it as much as possible with the packages in the internal directory.

norio-nomura commented:
> 3\. Overall, the code seems to have a tendency to overlap with existing code (does it support generative AI?).

The reason why it contains duplicate code is because it was written to work alone when this PR was not accepted.
I was already thinking about moving `xpc` and `vmnet` to lima.

Code-Hex commented on pkg/xpc/object.go:
```diff
@@ -0,0 +1,109 @@
+
+// Object represents a generic XPC object.
+//   - https://developer.apple.com/documentation/xpc/xpc_object_t?language=objc
+type Object interface {
```
@norio-nomura Thanks for your explanation. I do not wish users to interact with unsafe.Pointer as a feature provided by the library. The reason being that users may inadvertently utilise unintended library specifications, which are then treated as bugs requiring correction.

However, I understand your point, but I'd be very happy if we could find a solution that doesn't require exposing unsafe.Pointer.

norio-nomura committed:
Create `internal/osversion` from `osversion.go`

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

norio-nomura committed:
`xpc`: Deduplicate codes, add missing APIs, hide `unsafe.Pointer`

`xpc`: Use `internal/objc`

`xpc`: Add missing APIs, hide `unsafe.Pointer`, and unexport `xpcObject`

`xpc`: Change Fd from `int` to  `uintptr`

`xpc`: Remove calling `ReleaseOnCleanup()` in `xpcObject.retain()`

  Call `ReleaseOnCleanup()` explicitly if needed.

`xpc`: minor tweak

norio-nomura committed:
`vmnet`: Add `Interface` related APIs and `*FileAdapterForInterface`s

`vmnet`: Fix golangci-lint-v2 violations

`vmnet`: if iface.EnableVirtioHeader { packetSize += virtioNetHdrSize }

`vmnet`: Remove `object`

`vmnet`: Add doc comments to `*FileAdapterForInterface`s

`vmnet`: Refactor `*FileAdapterForInterface`s

- Remove written packet count from result of `WritePacketsTo*` in `PacketForwarder`
- Minimize differences between `pktDescsManager` and `msgHdrXArray`

`vmnet`: Handle `syscall.ENOBUFS` in `DatagramPacketForwarder.WritePacketsToConn`

`syscall.Sendmsg` may return `syscall.ENOBUFS` if there is not enough buffer set to the destination.

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

norio-nomura committed:
`internal/cgohandler`: Separate from `xpc` package

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

norio-nomura committed:
`vmnet_test.go`: Add `datagram*.FileAdapterForInterface` to test

Use `*FileAdapterForInterface` APIs in `TestVmnetSharedModeAllowsCommunicationBetweenMultipleVMs`:
- `datagram.FileAdapterForInterface`
- `datagramx.FileAdapterForInterface`
Since they are compatible with `vz.NewFileHandleNetworkDeviceAttachment`.

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

norio-nomura commented:
I think the approximately necessary APIs have been added.
Do I need to add tests to the xpc package?

norio-nomura committed:
`xpc`: Add unit tests

- Fix `NewArray` with empty slices.
- Change `New*` to return actual typed instance instead of `Object`

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

norio-nomura commented:
Unit tests added to `xpc`.
Previously, when I tried [`vmnet_network_configuration_add_port_forwarding_rule`](https://developer.apple.com/documentation/vmnet/vmnet_network_configuration_add_port_forwarding_rule(_:_:_:_:_:_:)?language=objc), it didn't work well, so [`vmnet_interface_add_ip_port_forwarding_rule`](https://developer.apple.com/documentation/vmnet/vmnet_interface_add_ip_port_forwarding_rule(_:_:_:_:_:_:_:)?language=objc) related to `Interface` has not been implemented, but I want to make it a separate PR to support them in the future.

norio-nomura committed:
`.clang-format`: Add `Language: Cpp`

- Remove `FIXME`s for "Configuration file(s) do(es) not support C++: ..."

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

norio-nomura committed:
`vmnet`: Refactor `*FileAdapterForInterface`s

- `StreamFileAdapterForInterface`:
  - Support partial read on `unix.Readv` in `readPacketsFromConn`

- `Datagram*FileAdapterForInterface`:
  - Add wait on `syscall.ENOBUFS` in `writePacketsToPacketConn`

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

Code-Hex reviewed (commented):
It's almost LGTM!

norio-nomura committed:
`xpc`: Change to delete `cgoReplyHandler` in `callReplyHandler`

To avoid crash when `Context` is canceled before receiving reply.

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

Code-Hex commented on vmnet/stream_darwin.go:
```diff
@@ -0,0 +1,236 @@
+	// Wait until 4-byte header is read
+	if _, err := conn.Read(v.backingBuffers[packetCount][:headerSize]); err != nil {
+		return 0, fmt.Errorf("conn.Read failed: %w", err)
+	}
```
I see. Thank you!

Code-Hex commented:
@norio-nomura  Sorry for the delay. Am I correct in understanding that this issue will take a little more time?
> The performance of FileAdaptorForInterface is not satisfactory, so I'm trying to rewrite it.

Code-Hex commented:
@norio-nomura Thank you for the explanation. I'm not in a hurry. I think the most important thing is to finish something great.

I am currently on a lot of business trips and have little time to write reviews, so I apologize for the delay in responding. Thank you in advance.

norio-nomura committed:
git mv

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

norio-nomura committed:
re-package

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

norio-nomura committed:
adaptor -> adapter

Signed-off-by: Norio Nomura <norio.nomura@gmail.com>

bubunyo commented on vmnet/vmnet_darwin.go:
```diff
@@ -0,0 +1,659 @@
+	if !subnet.Addr().Is4() {
+		return fmt.Errorf("subnet is not ipv4")
+	}
+	if !netip.MustParsePrefix("192.168.0.0/16").Overlaps(subnet) {
+		return fmt.Errorf("subnet %s is out of range", subnet.String())
+	}
```
Is there any specific reason for limiting to the 192.168.0.0/16 range only? Ideally users should be able to specify any RFC 1918 range, including 10.0.0.0/8 or 172.16.0.0/12 and it doesnt look like apple explicitly prevents this. 

```suggestion
rfc1918 := []netip.Prefix{
    netip.MustParsePrefix("10.0.0.0/8"),
    netip.MustParsePrefix("172.16.0.0/12"),
    netip.MustParsePrefix("192.168.0.0/16"),
}
allowed := false
for _, r := range rfc1918 {
    if r.Overlaps(subnet) {
        allowed = true
        break
    }
}
if !allowed {
    return fmt.Errorf("subnet %s is not in an RFC 1918 range", subnet.String())
}
```

nirs replied:
The docs say `a /24 under 192.168/16`.
https://developer.apple.com/documentation/vmnet/vmnet_network_configuration_create(_:_:)

bubunyo replied:
According to the doc, it is a default value. 
```
All other parameters are optional and have the following default value:
...
```

nirs replied:
Right, good chance that it supports the same private addresses as the older APIs.

bubunyo commented:
@norio-nomura any progress here, if you can list some of the changes you are looking for, i am happy to assist i making them and the adding then opening a pr to this.

s3rj1k commented:
Played with this a bit, specifically with

- vmnet.NewNetworkConfiguration(vmnet.SharedMode)
- vz.NewVmnetNetworkDeviceAttachment()
- netCfg.AddPortForwardingRule()

And I can confirm that at least this flow works fine and there is no need for root if below entitlements are used alongside with `--options runtime` to `codesign`

```
<dict>
    <key>com.apple.security.virtualization</key>
    <true/>
    <key>com.apple.security.cs.allow-jit</key>
    <true/>
    <key>com.apple.security.cs.allow-unsigned-executable-memory</key>
    <true/>
    <key>com.apple.security.cs.disable-library-validation</key>
    <true/>
</dict>
```

Hoping we get this merged before macOS 27.0 comes out, as it will also ship more new features.

---

One thing that I've noticed is that `AddDhcpReservation` kind of unreliable, guest can get different IP, this is fixable with custom cloud-init, but `AddDhcpReservation` itself enables `AddPortForwardingRule`, so I am happy this works.
